File: web_server/app/static/js/failed_orders.js
/**
 * Failed Orders Management System
 * @FEAT:immediate-order-execution @COMP:ui @TYPE:core
 *
 * Manages failed order display, retry, and removal functionality
 * Provides real-time updates and error handling for failed order operations
 *
 * Key Functions:
 * - loadFailedOrders(filters): Fetch and render failed orders
 * - renderFailedOrders(orders): Render table with order data
 * - retryFailedOrder(id): POST retry API call
 * - removeFailedOrder(id): DELETE API call
 * - filterFailedOrders(): Event handler for filter changes
 *
 * Dependencies:
 * - toast.js: showToast() for notifications
 * - base.html: CSRF token meta tag
 *
 * @see web_server/app/routes/failed_orders.py for API specification
 */

'use strict';

(function (window, document) {
    // Logger reference (graceful fallback)
    const log = window.logger || console;

    /**
     * Get CSRF token from meta tag
     * @returns {string} CSRF token value
     */
    function getCsrfToken() {
        const token = document.querySelector('meta[name="csrf-token"]')?.getAttribute('content') || '';
        if (!token) {
            log.warn('CSRF token not found in meta tag');
        }
        return token;
    }

    /**
     * Extract error message from API response
     * @param {object} data - API response data
     * @param {string} fallback - Fallback error message
     * @returns {string} Error message
     */
    function getErrorMessage(data, fallback = '알 수 없는 오류가 발생했습니다.') {
        if (!data || typeof data !== 'object') {
            return fallback;
        }

        // Direct error field
        if (typeof data.error === 'string') {
            return data.error;
        }

        // Error object with message field
        if (data.error && typeof data.error === 'object' && data.error.message) {
            return data.error.message;
        }

        // Message field
        if (typeof data.message === 'string') {
            return data.message;
        }

        return fallback;
    }

    /**
     * Escape HTML special characters to prevent XSS
     * @param {string} text - Text to escape
     * @returns {string} HTML-escaped text
     */
    function escapeHtml(text) {
        if (!text) return '';
        const div = document.createElement('div');
        div.textContent = text;
        return div.innerHTML;
    }

    /**
     * Fetch failed orders from API with optional filters
     * @param {object} filters - Filter parameters {strategy_account_id, symbol}
     * @returns {Promise<Array>} Array of failed order objects
     */
    async function loadFailedOrders(filters = {}) {
        try {
            // Build query parameters
            const params = new URLSearchParams();
            if (filters.strategy_account_id) {
                params.append('strategy_account_id', filters.strategy_account_id);
            }
            if (filters.symbol) {
                params.append('symbol', filters.symbol);
            }

            const queryString = params.toString();
            const url = `/api/failed-orders${queryString ? '?' + queryString : ''}`;

            log.debug('Loading failed orders from:', url);

            const response = await fetch(url, {
                method: 'GET',
                headers: {
                    'Content-Type': 'application/json',
                    'X-CSRFToken': getCsrfToken()
                }
            });

            if (!response.ok) {
                throw new Error(`HTTP ${response.status}: ${response.statusText}`);
            }

            const data = await response.json();

            if (data.success) {
                log.info(`Failed orders loaded: ${data.orders.length} orders`);
                renderFailedOrders(data.orders);
                return data.orders;
            } else {
                throw new Error(data.error || '주문 조회에 실패했습니다.');
            }

        } catch (error) {
            log.error('Failed to load failed orders:', error);
            const errorMsg = error.message || '실패한 주문을 불러오는데 실패했습니다.';
            showToast(errorMsg, 'error');
            renderFailedOrders([]);
            return [];
        }
    }

    /**
     * Render failed orders table with all order data
     * @param {Array} orders - Array of failed order objects
     */
    function renderFailedOrders(orders) {
        const tbody = document.getElementById('failed-orders-tbody');
        if (!tbody) {
            log.warn('Failed orders table body not found');
            return;
        }

        // Empty state
        if (!orders || orders.length === 0) {
            tbody.innerHTML = `
                <tr>
                    <td colspan="9" class="text-center" style="padding: 2rem;">
                        <svg style="width: 48px; height: 48px; margin: 0 auto 0.5rem; color: var(--text-muted);" fill="none" stroke="currentColor" viewBox="0 0 24 24">
                            <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M9 12l2 2 4-4m6 2a9 9 0 11-18 0 9 9 0 0118 0z"></path>
                        </svg>
                        <p style="color: var(--text-muted); margin: 0;">실패된 주문이 없습니다</p>
                    </td>
                </tr>
            `;
            return;
        }

        // Render each order row
        const rows = orders.map(order => {
            const sideBadgeClass = order.side === 'BUY' ? 'badge-buy' : 'badge-sell';
            const sideText = order.side === 'BUY' ? '매수' : '매도';

            // Retry count styling
            let retryCountHtml = '';
            if (order.retry_count > 0) {
                retryCountHtml = `<span class="retry-count-badge">${order.retry_count}/5</span>`;
            }

            // Disable retry button if max retries reached
            const isMaxRetriesReached = order.retry_count >= 5;
            const retryButtonDisabled = isMaxRetriesReached ? 'disabled' : '';
            const retryButtonTitle = isMaxRetriesReached ? '최대 재시도 횟수(5회)에 도달했습니다' : '이 주문을 재시도합니다';

            // Exchange error tooltip
            const exchangeErrorDisplay = order.exchange_error ? escapeHtml(order.exchange_error) : '없음';

            return `
                <tr data-failed-order-id="${order.id}">
                    <td>${escapeHtml(order.symbol)}</td>
                    <td><span class="${sideBadgeClass}">${sideText}</span></td>
                    <td>${escapeHtml(order.order_type)}</td>
                    <td>${parseFloat(order.quantity).toFixed(8)}</td>
                    <td>${order.price ? '$' + parseFloat(order.price).toFixed(4) : 'N/A'}</td>
                    <td>${escapeHtml(order.reason)}</td>
                    <td>
                        <span class="exchange-error-cell">
                            ${exchangeErrorDisplay}
                            ${order.exchange_error ? `<span class="exchange-error-tooltip">${escapeHtml(order.exchange_error)}</span>` : ''}
                        </span>
                    </td>
                    <td>${new Date(order.created_at).toLocaleString('ko-KR')}</td>
                    <td>
                        <button class="btn-retry"
                                onclick="window.retryFailedOrder(${order.id})"
                                title="${retryButtonTitle}"
                                ${retryButtonDisabled}>
                            재시도${retryCountHtml}
                        </button>
                        <button class="btn-remove"
                                onclick="window.removeFailedOrder(${order.id})"
                                title="이 주문을 영구 삭제합니다">
                            제거
                        </button>
                    </td>
                </tr>
            `;
        }).join('');

        tbody.innerHTML = rows;
        log.info(`Rendered ${orders.length} failed orders`);
    }

    /**
     * Retry a failed order via API
     * @param {number} failedOrderId - ID of the failed order to retry
     */
    async function retryFailedOrder(failedOrderId) {
        if (!confirm('이 주문을 재시도하시겠습니까?')) {
            return;
        }

        try {
            // Find and disable the button
            const button = document.querySelector(`tr[data-failed-order-id="${failedOrderId}"] .btn-retry`);
            if (button) {
                const originalText = button.innerHTML;
                button.disabled = true;
                button.innerHTML = '처리 중...';

                try {
                    const response = await fetch(`/api/failed-orders/${failedOrderId}/retry`, {
                        method: 'POST',
                        headers: {
                            'Content-Type': 'application/json',
                            'X-CSRFToken': getCsrfToken()
                        }
                    });

                    const data = await response.json();

                    if (data.success) {
                        const message = `재시도 성공: ${data.order_id}`;
                        log.info(message);
                        showToast(message, 'success');
                        // Reload to reflect updated retry count
                        await loadFailedOrders();
                    } else {
                        const errorMsg = getErrorMessage(data, '재시도 중 오류가 발생했습니다.');

                        // Handle specific error cases
                        if (response.status === 400 && data.error.includes('최대 재시도')) {
                            log.warn('Max retry attempts exceeded:', errorMsg);
                        }

                        showToast(errorMsg, 'error');
                        await loadFailedOrders();
                    }

                } finally {
                    // Restore button state
                    button.disabled = false;
                    button.innerHTML = originalText;
                }
            }

        } catch (error) {
            log.error('Failed to retry order:', error);
            const errorMsg = error.message || '재시도 중 오류가 발생했습니다.';
            showToast(errorMsg, 'error');
        }
    }

    /**
     * Remove a failed order via API
     * @param {number} failedOrderId - ID of the failed order to remove
     */
    async function removeFailedOrder(failedOrderId) {
        if (!confirm('이 주문을 영구 삭제하시겠습니까?')) {
            return;
        }

        try {
            // Find and disable the button
            const button = document.querySelector(`tr[data-failed-order-id="${failedOrderId}"] .btn-remove`);
            if (button) {
                const originalText = button.innerHTML;
                button.disabled = true;
                button.innerHTML = '삭제 중...';

                try {
                    const response = await fetch(`/api/failed-orders/${failedOrderId}`, {
                        method: 'DELETE',
                        headers: {
                            'Content-Type': 'application/json',
                            'X-CSRFToken': getCsrfToken()
                        }
                    });

                    const data = await response.json();

                    if (data.success) {
                        log.info('Failed order removed successfully:', failedOrderId);
                        showToast('주문이 삭제되었습니다', 'success');
                        // Reload to reflect deletion
                        await loadFailedOrders();
                    } else {
                        const errorMsg = getErrorMessage(data, '삭제 중 오류가 발생했습니다.');

                        // Handle specific error cases
                        if (response.status === 403) {
                            log.warn('Permission denied:', errorMsg);
                        } else if (response.status === 404) {
                            log.warn('Order not found:', errorMsg);
                        }

                        showToast(errorMsg, 'error');
                        await loadFailedOrders();
                    }

                } finally {
                    // Restore button state
                    button.disabled = false;
                    button.innerHTML = originalText;
                }
            }

        } catch (error) {
            log.error('Failed to remove order:', error);
            const errorMsg = error.message || '삭제 중 오류가 발생했습니다.';
            showToast(errorMsg, 'error');
        }
    }

    /**
     * Event handler for filter changes
     * Collects current filter values and triggers load
     */
    function filterFailedOrders() {
        const strategyAccountId = document.getElementById('filter-strategy')?.value || '';
        const symbol = document.getElementById('filter-symbol')?.value || '';

        const filters = {};
        if (strategyAccountId) {
            filters.strategy_account_id = parseInt(strategyAccountId, 10);
        }
        if (symbol) {
            filters.symbol = symbol;
        }

        log.debug('Applying filters:', filters);
        loadFailedOrders(filters);
    }

    /**
     * Initialize event listeners
     */
    function initializeEventListeners() {
        // Search button click handler
        const searchButton = document.querySelector('.failed-orders-section .filters button');
        if (searchButton) {
            searchButton.addEventListener('click', filterFailedOrders);
            log.debug('Search button listener attached');
        }

        // Enter key on filter inputs
        const strategySelect = document.getElementById('filter-strategy');
        const symbolInput = document.getElementById('filter-symbol');

        if (symbolInput) {
            symbolInput.addEventListener('keypress', (e) => {
                if (e.key === 'Enter') {
                    filterFailedOrders();
                }
            });
            log.debug('Symbol input listener attached');
        }

        if (strategySelect) {
            strategySelect.addEventListener('change', filterFailedOrders);
            log.debug('Strategy select listener attached');
        }
    }

    /**
     * Initialize the failed orders module
     */
    function initialize() {
        log.info('Initializing failed orders module');

        // Check if failed orders section exists
        const section = document.querySelector('.failed-orders-section');
        if (!section) {
            log.warn('Failed orders section not found in DOM');
            return;
        }

        // Initialize event listeners
        initializeEventListeners();

        // Load initial data
        loadFailedOrders();

        log.info('Failed orders module initialized successfully');
    }

    // Register global functions for inline onclick handlers
    window.retryFailedOrder = retryFailedOrder;
    window.removeFailedOrder = removeFailedOrder;
    window.filterFailedOrders = filterFailedOrders;

    // Initialize when DOM is ready
    document.addEventListener('DOMContentLoaded', initialize);

})(window, document);

File: web_server/app/routes/failed_orders.py
"""
실패한 주문 관리 API 엔드포인트

@FEAT:immediate-order-execution @COMP:route @TYPE:core

사용자가 웹 UI에서 실패한 주문을 조회, 재시도, 제거할 수 있도록
RESTful API 엔드포인트를 제공합니다.

엔드포인트:
- GET /api/failed-orders: 목록 조회 (필터: strategy_account_id, symbol)
- POST /api/failed-orders/<int:failed_order_id>/retry: 재시도
- DELETE /api/failed-orders/<int:failed_order_id>: 제거

권한 검증:
- 모든 엔드포인트는 @login_required 적용
- FailedOrder → StrategyAccount → Strategy → User 체인 확인
- 현재 사용자 소유의 FailedOrder만 접근 가능
"""
from flask import Blueprint, jsonify, request, current_app
from flask_login import login_required, current_user
from app.models import FailedOrder, StrategyAccount, Account
from app.services.trading.failed_order_manager import failed_order_manager
from sqlalchemy.orm import joinedload

# @FEAT:immediate-order-execution @COMP:route @TYPE:core
failed_orders_bp = Blueprint('failed_orders', __name__)


# @FEAT:immediate-order-execution @COMP:route @TYPE:core
@failed_orders_bp.route('/api/failed-orders', methods=['GET'])
@login_required
def get_failed_orders():
    """
    실패한 주문 목록 조회 (status='pending_retry'만 반환)

    Query parameters:
    - strategy_account_id (int, optional): 전략 계정 ID 필터
    - symbol (str, optional): 심볼 필터 (예: "BTC/USDT")

    Returns:
        200: {
            'success': True,
            'orders': [
                {
                    'id': 1,
                    'strategy_account_id': 1,
                    'symbol': 'BTC/USDT',
                    'side': 'BUY',
                    'order_type': 'LIMIT',
                    'quantity': '0.1',
                    'price': '50000',
                    'stop_price': None,
                    'market_type': 'SPOT',
                    'reason': '잔고 부족',
                    'exchange_error': 'Insufficient balance',
                    'retry_count': 0,
                    'status': 'pending_retry',
                    'created_at': '2025-10-29T10:00:00',
                    'updated_at': '2025-10-29T10:00:00'
                }
            ]
        }
        401: {'success': False, 'error': 'Unauthorized'}
        500: {'success': False, 'error': 'Internal server error'}
    """
    try:
        # Query parameters (optional)
        strategy_account_id = request.args.get('strategy_account_id', type=int)
        symbol = request.args.get('symbol', type=str)

        # 1. 현재 사용자의 모든 StrategyAccount ID 조회
        # StrategyAccount → Strategy → User 체인 확인
        user_strategy_account_ids = (
            StrategyAccount.query
            .join(StrategyAccount.strategy)
            .filter_by(user_id=current_user.id)
            .with_entities(StrategyAccount.id)
            .all()
        )
        user_strategy_account_ids = [sa_id[0] for sa_id in user_strategy_account_ids]

        if not user_strategy_account_ids:
            # 사용자가 소유한 StrategyAccount가 없음
            return jsonify({'success': True, 'orders': []})

        # 2. FailedOrderManager를 통한 조회
        failed_orders = failed_order_manager.get_failed_orders(
            strategy_account_id=strategy_account_id,
            symbol=symbol
        )

        # 3. 사용자 소유 StrategyAccount에 속한 주문만 필터링
        filtered_orders = [
            fo for fo in failed_orders
            if fo.strategy_account_id in user_strategy_account_ids
        ]

        # 4. 응답 데이터 구성
        orders = []
        for fo in filtered_orders:
            orders.append({
                'id': fo.id,
                'strategy_account_id': fo.strategy_account_id,
                'symbol': fo.symbol,
                'side': fo.side,
                'order_type': fo.order_type,
                'quantity': str(fo.quantity),  # Decimal → str (JSON 직렬화)
                'price': str(fo.price) if fo.price else None,
                'stop_price': str(fo.stop_price) if fo.stop_price else None,
                'market_type': fo.market_type,
                'reason': fo.reason,
                'exchange_error': fo.exchange_error,
                'retry_count': fo.retry_count,
                'status': fo.status,
                'created_at': fo.created_at.isoformat() if fo.created_at else None,
                'updated_at': fo.updated_at.isoformat() if fo.updated_at else None
            })

        return jsonify({'success': True, 'orders': orders})

    except Exception as e:
        current_app.logger.error(f'실패 주문 조회 실패: {str(e)}')
        return jsonify({'success': False, 'error': 'Internal server error'}), 500


# @FEAT:immediate-order-execution @COMP:route @TYPE:core
@failed_orders_bp.route('/api/failed-orders/<int:failed_order_id>/retry', methods=['POST'])
@login_required
def retry_failed_order(failed_order_id):
    """
    실패한 주문 재시도

    Args:
        failed_order_id (int): 재시도할 FailedOrder의 ID

    Returns:
        200: {
            'success': True,
            'order_id': 'BINANCE_ORDER_12345',
            'message': '재시도 성공'
        }
        400: {
            'success': False,
            'error': '최대 재시도 횟수(5회)를 초과했습니다'
        }
        403: {'success': False, 'error': '권한이 없습니다'}
        404: {'success': False, 'error': '주문을 찾을 수 없습니다'}
        500: {'success': False, 'error': 'Internal server error'}
    """
    try:
        # 1. FailedOrder 조회 (존재 여부 확인)
        failed_order = (
            FailedOrder.query
            .options(
                joinedload(FailedOrder.strategy_account)
                .joinedload(StrategyAccount.strategy)
            )
            .filter_by(id=failed_order_id)
            .first()
        )

        if not failed_order:
            return jsonify({'success': False, 'error': '주문을 찾을 수 없습니다'}), 404

        # 2. 권한 검증: FailedOrder → StrategyAccount → Strategy → User 체인
        if not failed_order.strategy_account or not failed_order.strategy_account.strategy:
            current_app.logger.error(
                f'FailedOrder {failed_order_id} 권한 체크 실패: '
                f'strategy_account 또는 strategy 누락'
            )
            return jsonify({'success': False, 'error': '권한이 없습니다'}), 403

        if failed_order.strategy_account.strategy.user_id != current_user.id:
            return jsonify({'success': False, 'error': '권한이 없습니다'}), 403

        # 3. FailedOrderManager를 통한 재시도
        result = failed_order_manager.retry_failed_order(failed_order_id)

        if result['success']:
            return jsonify({
                'success': True,
                'order_id': result.get('order_id'),
                'message': '재시도 성공'
            })
        else:
            # 실패 원인에 따라 적절한 HTTP 상태 코드 반환
            error_msg = result.get('error', 'Unknown error')

            # 최대 재시도 횟수 초과
            if 'Maximum retry count exceeded' in error_msg:
                return jsonify({'success': False, 'error': error_msg}), 400

            # 기타 실패 (잔고 부족, API 오류 등)
            return jsonify({'success': False, 'error': error_msg}), 400

    except Exception as e:
        current_app.logger.error(f'실패 주문 재시도 실패: {str(e)}')
        return jsonify({'success': False, 'error': 'Internal server error'}), 500


# @FEAT:immediate-order-execution @COMP:route @TYPE:core
@failed_orders_bp.route('/api/failed-orders/<int:failed_order_id>', methods=['DELETE'])
@login_required
def delete_failed_order(failed_order_id):
    """
    실패한 주문 제거 (사용자 수동 제거)

    Args:
        failed_order_id (int): 제거할 FailedOrder의 ID

    Returns:
        200: {
            'success': True,
            'message': '주문이 제거되었습니다'
        }
        403: {'success': False, 'error': '권한이 없습니다'}
        404: {'success': False, 'error': '주문을 찾을 수 없습니다'}
        500: {'success': False, 'error': 'Internal server error'}

    Notes:
        - CSRF Protection: Flask-WTF CSRFProtect가 자동으로 DELETE 요청 검증
        - JavaScript에서 X-CSRFToken 헤더 전송 필요 (Phase 7에서 구현)
    """
    try:
        # 1. FailedOrder 조회 (존재 여부 확인)
        failed_order = (
            FailedOrder.query
            .options(
                joinedload(FailedOrder.strategy_account)
                .joinedload(StrategyAccount.strategy)
            )
            .filter_by(id=failed_order_id)
            .first()
        )

        if not failed_order:
            return jsonify({'success': False, 'error': '주문을 찾을 수 없습니다'}), 404

        # 2. 권한 검증: FailedOrder → StrategyAccount → Strategy → User 체인
        if not failed_order.strategy_account or not failed_order.strategy_account.strategy:
            current_app.logger.error(
                f'FailedOrder {failed_order_id} 권한 체크 실패: '
                f'strategy_account 또는 strategy 누락'
            )
            return jsonify({'success': False, 'error': '권한이 없습니다'}), 403

        if failed_order.strategy_account.strategy.user_id != current_user.id:
            return jsonify({'success': False, 'error': '권한이 없습니다'}), 403

        # 3. FailedOrderManager를 통한 제거
        success = failed_order_manager.remove_failed_order(failed_order_id)

        if success:
            return jsonify({
                'success': True,
                'message': '주문이 제거되었습니다'
            })
        else:
            # remove_failed_order는 FailedOrder를 찾지 못하면 False 반환
            # 하지만 위에서 이미 존재 여부를 확인했으므로 이 경로는 드물게 발생
            return jsonify({'success': False, 'error': '주문 제거 실패'}), 500

    except Exception as e:
        current_app.logger.error(f'실패 주문 제거 실패: {str(e)}')
        return jsonify({'success': False, 'error': 'Internal server error'}), 500
